buildSrc/subprojects/parser/src/test/java/org/apache/calcite/test/DialectGenerateTest.java

```diff
@@ -0,0 +1,158 @@
+    // Adds the path /intermediate/testDialect/ to the end of rootPath.
+    Path dialectPath = rootPath.resolve(Paths.get("intermediate", "testDialect"));
+    File rootFile = rootPath.toFile();
+    File dialectFile = dialectPath.toFile();
+    return new DialectGenerate(dialectFile, rootFile, "");
+  }
+
+  /**
+   * Returns the expected extraction results after calling DialectGenerate.extractFunctions
+   * onto the parsingTest directory structure.
+   */
+  private Map<String, String> getExpectedExtractionResults() {
+    String foo = "void foo() :\n"
+      + "{\n"
+      + "    String x = \" '}' \";\n"
+      + "}\n"
+      + "{\n"
+      + "    // overidden by testDialect\n"
+      + "}";
+    String bar = "void bar  (   )  : {} {}";
+    String baz = "void baz() : {x}\n"
+      + "{\n"
+      + "    // overridden by intermediate\n"
+      + "}";
+    String qux = "void qux(int arg1, int arg2) :\n"
+      + "{\n"
+      + "    String x = \" } \"\n"
+      + "}\n"
+      + "{\n"
+      + "    // Below is a }\n"
+      + "}";
+    String quux = "void quux(int arg1,\n"
+      + "    int arg2\n"
+      + ") :\n"
+      + "{\n"
+      + "    char x = ' } '\n"
+      + "    String y;\n"
+      + "}\n"
+      + "{\n"
+      + "    /* All invalid curly braces:\n"
+      + "    }\n"
+      + "    // }\n"
+      + "    \" } \"\n"
+      + "    ' } '\n"
+      + "    */\n"
+      + "    <TOKEN> {\n"
+      + "        y = \" { } } \"\n"
+      + "    }\n\n"
+      + "    // Not a string: \"\n"
+      + "}";
+    return new LinkedHashMap<String, String>() {{
+      put("foo", foo);
+      put("bar", bar);
+      put("baz", baz);
+      put("qux", qux);
+      put("quux", quux);
+    }};
+  }
+
+  private Queue<String> getTokens(String str) {
+    return new LinkedList(Arrays.asList(DialectGenerate.tokenizerPattern.split(str)));
+  }
+
+  private void assertCurlyParserSucceeds(String input) {
+    CurlyParser parser = new CurlyParser();
+    Queue<String> tokens = getTokens(input);
+    boolean doneParsing = false;
+    while (!tokens.isEmpty()) {
+      if (doneParsing) {
+        assertTrue(false, "Failed to parse curly block");
+      }
+      doneParsing = parser.parseToken(tokens.poll());
+    }
+    assertTrue(doneParsing);
+  }
+
+  @Test public void testCurlyParserParsesEmpty() {
+    String input = "{  }";
+    assertCurlyParserSucceeds(input);
+  }
+
+  @Test public void testCurlyParserParsesString() {
+    String input = "{ \" } \"  }";
+    assertCurlyParserSucceeds(input);
+  }
+
+  @Test public void testCurlyParserParsesCharacter() {
+    String input = "{ ' } '  }";
+    assertCurlyParserSucceeds(input);
+  }
+
+  @Test public void testCurlyParserParsesSingleComment() {
+    String input = "{ //  } \n  }";
+    assertCurlyParserSucceeds(input);
+  }
+
+  @Test public void testCurlyParserParsesMultiComment() {
+    String input = "{ /*  } */  }";
+    assertCurlyParserSucceeds(input);
+  }
+
+  @Test public void testExtractFunctions() {
```

Comment:
It could be good to setup a test for when the input is malformed or missing a brace to show it should throw an IllegalStateException when it fails

Reply:
The exception is currently being caught in the try-catch block so I don't think that would be possible.

Reply:
Ah I see. So would it just skip over that ftl file if it's malformed? Should the build just fail instead and throw the error rather than skipping over the logic if that's the case?

Reply:
Currently what happens is it would add what was parsed so far to the function map so the resultant generated parserImpls.ftl will be wrong. I think it would make more sense to break the build there but I would like @dasch-1 to give some input.

Reply:
Actually never-mind, you are right. If an error is encountered within a .ftl file, everything after that error will not be parsed.